### PR TITLE
BFD-1691: Store performance statistics in JSON format to S3 or file if specified

### PR DESCRIPTION
**Pull request by malessi**
**JIRA Ticket:**
[BFD-1691](https://jira.cms.gov/browse/BFD-1691)

**User Story or Bug Summary:**

As a BFD engineer, I want to capture and store performance performance metrics on configured runs, so that I can store performance metrics which will be reviewed to assess the performance of the BFD as new functionality gets added.

---

### What Does This PR Do?

This PR:

* Adds a new file, `stats.py`, that includes several new classes and enums:
  * A new class, `AggregatedStats`, which exposes a computed property `all_stats` that returns a dictionary of the aggregated stats of all endpoints along with some extra metadata (timestamp, runtime, etc)
  * Two writer classes, `StatsJsonFileWriter` and `StatsJsonS3Writer`. These classes each expose a `write()` method that writes the stats from `AggregatedStats.all_stats` to either a local file or an S3 Bucket (using `boto3`)
    * Which writer is used is controlled by the user-supplied `--storeStats` command-line argument explained below
    * `StatsJsonS3Writer` writes to a pre-determined S3 path that follows the BFD Insights data convention: `<S3_BUCKET_PROVIDED>/databases/bfd/test_performance_stats/env=<ENV_PROVIDED>/tag=<TAG_PROVIDED>/<TIMESTAMP>.json`
    * `StatsJsonFileWriter` writes to a user-provided _parent path_, but the filename always follows the format: `<ENV_PROVIDED>-<TAG_PROVIDED>-<TIMESTAMP>.json`
  * Some new `dataclass`s , `StatsStorageConfig`, `StatsFileStorageConfig` and `StatsS3StorageConfig`, to hold state about where aggregated stats should be stored
  * Some `enum`s, `StatsStorageType` and `StatsEnvironment`, to encode valid types of storage (to a local file \[`FILE`\] or to an S3 Bucket \[`S3`\]) and valid running environments (`TEST` or `PROD`)
* Adds a new command-line argument, `--storeStats`, which requires the user provide a specially-formatted "arg string" (in the format `<STORAGE_TYPE>:<RUNNING_ENVIRONMENT>:<TAG>:<PATH_OR_BUCKET>`) that tells the code where to store performance statistics (file or S3) and how to store them (path or bucket)
  * Documentation was added to the README explaining the argument format in-depth, as well
* Adds some code to `config.py` that modifies PyYAML's dumper and loader to understand how to serialize and deserialize `StatsStorageConfig` objects to the string format used by the `--storeStats` command-line argument (`<STORAGE_TYPE>:<RUNNING_ENVIRONMENT>:<TAG>:<PATH_OR_BUCKET>`)
* Introduces a run-once "teardown" method to `bfd_user_base.py` which listens on Locust's `test_stop` event to write performance statistics to either S3 or to a local file depending on the value of `--storeStats`
* Includes a few small fixes:
  * A fix for a `TypeError` being raised when the default `config.yml` is empty
  * A fix for a `TypeError` being raised due to Locust requiring the `LOCUST_RESET_STATS` environment variable be a `str`, not a `bool`

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Are the chosen S3 path and BFD Insights database name (`test_performance_stats`) applicable? 
  * Are the chosen partitions (`env` and `tag`) applicable as well?
  * Do these need to be named differently?
  * Should they be configurable?
* Are the correct performance statistics being collected?
  * Should machine information also be collected?
  * Are there any other metrics that would be useful to collect?
* Is there any additional documentation required? 
* Are the code changes and additions understandable?
* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A** 

    * Does this PR add any new software dependencies? 
      * [x] Yes
      * [ ] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by StewGoin — COMMENTED** [review]
I see the new dependency on boto3 which is a pretty capable but also ought to be well looked after library. I think we can include it without additional SIA as long as we ensure we're using it according to best practices / developer guide recommendations from AWS and others.

I am curious about the storage to S3 / file -- if going to S3, is that a protected / encrypted bucket? And are there any concerns / thoughts around if these tests and performance statistics could be an avenue for accidental leakage of sensitive information?

**Comment by malessi** [comment]
> I see the new dependency on boto3 which is a pretty capable but also ought to be well looked after library. I think we can include it without additional SIA as long as we ensure we're using it according to best practices / developer guide recommendations from AWS and others.
> 
> I am curious about the storage to S3 / file -- if going to S3, is that a protected / encrypted bucket? And are there any concerns / thoughts around if these tests and performance statistics could be an avenue for accidental leakage of sensitive information?

Sounds good as far as `boto3`. The only usage of it in this set of changes is to upload to S3 -- I do not expect we will be doing much more with it in the future. Also, credentials are _not_ specified in the code.

With these changes the user can specify the bucket that the JSON should be stored in. For now, that bucket will almost always be an existing BFD Insights bucket that appears to have the proper security controls (encryption, non-public, etc). As far as the actual statistics themselves, I don't believe there is anything sensitive exposed by them by default _unless_ the user specifies a sensitive `tag`. Here's an example of what the statistics look like:

```json
{
    "timestamp": 1653421860,
    "tag": "some_tag",
    "environment": "TEST",
    "statsResetAfterSpawn": false,
    "numUsers": 5,
    "usersPerSecond": 5.0,
    "runtime": 10.85379672050476,
    "endpoints": [
        {
            "name": "Example Test",
            "requestMethod": "GET",
            "numRequests": 657,
            "numFailures": 0,
            "medianResponseTime": 39,
            "avgResponseTime": 40.11252364535764,
            "minResponseTime": 34.21466700000009,
            "maxResponseTime": 173.93270800000005,
            "avgContentLength": 1256.0,
            "totalRequestsPerSecond": 60.53181360572284,
            "totalFailuresPerSecond": 0.0,
            "50%": 39,
            "66%": 40,
            "75%": 41,
            "80%": 42,
            "90%": 43,
            "95%": 44,
            "98%": 46,
            "99%": 47,
            "99.9%": 170,
            "99.99%": 170,
            "100%": 170
        },
        {
            "name": "Example Test2",
            "requestMethod": "GET",
            "numRequests": 685,
            "numFailures": 0,
            "medianResponseTime": 39,
            "avgResponseTime": 40.186863131386815,
            "minResponseTime": 34.886917000000125,
            "maxResponseTime": 174.52120799999997,
            "avgContentLength": 1256.0,
            "totalRequestsPerSecond": 63.11155604249642,
            "totalFailuresPerSecond": 0.0,
            "50%": 39,
            "66%": 40,
            "75%": 41,
            "80%": 42,
            "90%": 44,
            "95%": 45,
            "98%": 47,
            "99%": 48,
            "99.9%": 170,
            "99.99%": 170,
            "100%": 170
        }
    ]
}
```

File: ops/ccs-ops-misc/load_test/common/stats.py
Comment (lsmitchell) [inline ops/ccs-ops-misc/load_test/common/stats.py:2]: Maybe mention the primary purpose of the class here.

File: ops/ccs-ops-misc/load_test/README.md
Comment (malessi) [reply ops/ccs-ops-misc/load_test/README.md:122]: Agreed; I'll update the documentation here and add another bullet point about `file` mainly being for debugging and local testing purposes.

File: ops/ccs-ops-misc/load_test/common/stats.py
Comment (malessi) [reply ops/ccs-ops-misc/load_test/common/stats.py:2]: There's documentation for the classes within this file already, so I'll expand upon this comment and explain what code should live in `stats.py`.

File: ops/ccs-ops-misc/load_test/common/test_setup.py
Comment (malessi) [reply ops/ccs-ops-misc/load_test/common/test_setup.py:40]: Yeah, I was thinking that as I was writing it. Tbh, I feel like _all_ of the loading functions in this file should all be moved to `config.py`, not just this single function.

I think I'll go ahead with that refactor and move everything that loads from the config into `config.py` so that `config.py` is all the config-related stuff and `test_setup.py` is just meant for code that sets up the Locust environment and related things.

File: ops/ccs-ops-misc/load_test/runtests.py
Comment (andrew-kramer) [inline ops/ccs-ops-misc/load_test/runtests.py:14]: Is `StatsEnvironment` used here?

File: ops/ccs-ops-misc/load_test/common/stats.py
Comment (andrew-kramer) [inline ops/ccs-ops-misc/load_test/common/stats.py:136]: super-nit: Parameter is named `locust_env`

File: ops/ccs-ops-misc/load_test/common/stats.py
Comment (andrew-kramer) [inline ops/ccs-ops-misc/load_test/common/stats.py:222]: Also worth noting: if the test terminates early, this method will be more accurate than the user-provided runtime.

**Review by andrew-kramer — APPROVED** [review]
Looks great! This will be a big improvement.

Couple _very_ small things, but certainly nothing that would block.

File: ops/ccs-ops-misc/load_test/common/bfd_user_base.py
Comment (malessi) [reply ops/ccs-ops-misc/load_test/common/bfd_user_base.py:185]: That was originally where I had put this code; unfortunately, `on_stop` is called for _each_ user in the environment. So, if the user count is greater than 1, the statistics will be written to disk/S3 more than once which is not what we want.

Maybe I'll rename `teardown` to `one_time_teardown` to match conventions from other testing frameworks like NUnit and the like, just so it's a little more clear.

File: ops/ccs-ops-misc/load_test/common/stats.py
Comment (malessi) [reply ops/ccs-ops-misc/load_test/common/stats.py:22]: It's not super necessary and might be a little confusing, but I'm basically re-exporting Locust's `PERCENTILES_TO_REPORT` in `stats.py` so that consumers of `stats.py` (our code) don't need to import from another file if they want to use the default percentiles. Probably more confusing than it's worth, so I can definitely remove it if we feel it's a bit of a code smell.

File: ops/ccs-ops-misc/load_test/runtests.py
Comment (malessi) [reply ops/ccs-ops-misc/load_test/runtests.py:14]: Good catch! A previous iteration of these changes had multiple command-line arguments with one specifying the environment and the code in `runtests.py` handling that argument used `StatsEnvironment`.

I'll remove the unused import.

File: ops/ccs-ops-misc/load_test/common/stats.py
Comment (malessi) [reply ops/ccs-ops-misc/load_test/common/stats.py:136]: Ah, I didn't catch that one when I refactored that parameter. I'll update it to `locust_env`.

File: ops/ccs-ops-misc/load_test/common/bfd_user_base.py
Comment (andrew-kramer) [reply ops/ccs-ops-misc/load_test/common/bfd_user_base.py:185]: OK, yeah, I see.

I _think_ you could restrict it by `if not LOCUST_WORKER_NUM` , which should catch the case of `None` (there is no multi-threading) and `0` (there is multi-threading but this is the owner node). Without testing it, I'm not totally sure about that approach from this particular function. We have helper functions to determine this, but again kinda the same issue: they may not work during teardown.

I'm definitely onboard with just changing the name to match other conventions.

File: ops/ccs-ops-misc/load_test/common/stats.py
Comment (andrew-kramer) [reply ops/ccs-ops-misc/load_test/common/stats.py:22]: Makes sense. I should have realized that. I'd say leave it alone, maybe add a quick comment?

**Review by lsmitchell — APPROVED** [review]
Thanks for the fixes!

**Review by andrew-kramer — APPROVED** [review]
LGTM! Great work!

**Review by christrotteradhoc — APPROVED** [review]
Great job. I think this one has been thoroughly reviewed by the team and I'm approving based on their reviews. After this is merged, I think it is worth a small story to move the load_test into the apps/utils so that the codeowners will be more appropriately set for these files.

**Pull request merged**
